### PR TITLE
Fix building against OpenSSL 1.1.0

### PR DESCRIPTION
**Pull request by andhe**

### CODE REVIEW

**Comment by andhe** [comment]
This has only been compile-tested. Please review carefully.

File: cipher.c [hunk 1]
```diff
@@ -147,36 +145,35 @@ int cipher_rsa_encrypt(const char *plaintext,
 
 char *cipher_aes_decrypt(const unsigned char *ciphertext, size_t len, const unsigned char key[KDF_HASH_LEN])
 {
-	EVP_CIPHER_CTX ctx;
+	EVP_CIPHER_CTX *ctx = EVP_CIPHER_CTX_new();
```
Comment (bcopeland) [inline cipher.c]: Surely this is missing a null check?

File: config.c [hunk 2]
```diff
@@ -353,23 +352,21 @@ static size_t encrypt_buffer(const char *buffer, size_t in_len, unsigned const c
 	return len;
 
 error:
-	EVP_CIPHER_CTX_cleanup(&ctx);
+	EVP_CIPHER_CTX_cleanup(ctx);
```
Comment (bcopeland) [inline config.c]: probably free is needed here.

File: config.c [hunk 3]
```diff
@@ -353,23 +352,21 @@ static size_t encrypt_buffer(const char *buffer, size_t in_len, unsigned const c
...
 	free(ciphertext);
 	die("Failed to encrypt data.");
 
 }
 
 static size_t decrypt_buffer(const unsigned char *buffer, size_t in_len, unsigned const char key[KDF_HASH_LEN], unsigned char **out)
 {
-	EVP_CIPHER_CTX ctx;
+	EVP_CIPHER_CTX *ctx = EVP_CIPHER_CTX_new();
```
Comment (bcopeland) [inline config.c]: missing null check

File: config.c [hunk 4]
```diff
@@ -379,20 +376,20 @@ static size_t decrypt_buffer(const unsigned char *buffer, size_t in_len, unsigne
...
 		goto error;
 	len += out_len;
-	EVP_CIPHER_CTX_cleanup(&ctx);
+	EVP_CIPHER_CTX_free(ctx);
 	*out = plaintext;
 	return len;
 
 error:
-	EVP_CIPHER_CTX_cleanup(&ctx);
+	EVP_CIPHER_CTX_cleanup(ctx);
```
Comment (bcopeland) [inline config.c]: _free... and so on

File: pbkdf2.c [hunk 5]
```diff
@@ -40,7 +40,7 @@ int fallback_pkcs5_pbkdf2_hmac(const char *pass, size_t pass_len,
 	const unsigned char *salt, size_t salt_len, unsigned int iterations,
 	const EVP_MD *digest, size_t key_len, unsigned char *output)
 {
-	HMAC_CTX ctx;
+	HMAC_CTX *ctx = HMAC_CTX_new();
```
Comment (sebastianas) [inline pbkdf2.c]: NULL check

**Comment by andhe** [comment]
Hello @bcopeland and @sebastianas 
Thanks for your reviews. I would very much appreciate if someone who actually uses lastpass would take it from here, not only fix the review comments but also actually runtime test the application with the changes applied. (I'm not a lastpass user myself.)

**Comment by bcopeland** [comment]
Ok, I'll take over this.  I'll close the PR and redo it.

**Pull request closed without merging**